File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = set()
        nums.sort()
        n = len(nums)
        for i in range(n):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            left = i+1
            right = n - 1
            while left < right:
                cur = nums[left] + nums[right]
                if cur == -nums[i]:
                    result.add((nums[i], nums[left], nums[right]))
                if cur > -nums[i]:
                    right -= 1
                else:
                    left += 1
        return list(result)

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        Set<List<Integer>> res = new HashSet<>();
        for (int i = 0; i < n; i++) {
            int left = i+1;
            int right = n-1;
            int target = -nums[i];
            while (left < right) {
                if (nums[left] + nums[right] == target) {
                    List<Integer> arr = new ArrayList<>();
                    arr.add(-target);
                    arr.add(nums[left]);
                    arr.add(nums[right]);
                    res.add(arr);
                }
                if (nums[left] + nums[right] > target) {
                    right--;
                } else {
                    left++;
                }
            }
        }
        return new ArrayList<>(res);
    }
}